src/QsCompiler/Transformations/Monomorphization.cs

```diff
@@ -6,47 +6,187 @@
+    }
+
+    #region ResolveGenerics
+
     public class ResolveGenericsSyntax :
-        SyntaxTreeTransformation<ScopeTransformation<ResolveGenericsExpression>>
+        SyntaxTreeTransformation<NoScopeTransformations>
     {
-        public static IEnumerable<QsNamespace> Apply(IEnumerable<QsNamespace> namespaces)
+        ImmutableDictionary<NonNullable<string>, IEnumerable<QsCallable>> NamespaceCallables;
+
+        public static IEnumerable<QsNamespace> Apply(IEnumerable<QsNamespace> namespaces, QsQualifiedName entryPointName)
         {
             if (namespaces == null || namespaces.Contains(null)) throw new ArgumentNullException(nameof(namespaces));
 
-            // Get list of generics
-            List<QsQualifiedName> generics = new List<QsQualifiedName>();
-            var Callables = namespaces.GlobalCallableResolutions();
-            foreach (var callable in Callables)
+            var globals = namespaces.GlobalCallableResolutions();
+
+            var entryPoints = globals
+                .Where(call => // TODO: get list of entry points
+                    call.Key.Namespace.Value == entryPointName.Namespace.Value &&
+                    call.Key.Name.Value == entryPointName.Name.Value)
+                .Select(call => new Request
+                {
+                    originalName = call.Key,
+                    typeResolutions = null,
+                    concreteName = call.Key
+                });
+
+            var requests = new Stack<Request>(entryPoints);
+            var responses = new List<Response>();
+
+            while(requests.Count() > 0)
             {
-                if (callable.Value.Signature.TypeParameters.Any())
+                Request currentRequest = requests.Pop();
+
+                // If there is a call to an unknown callable, throw exception
+                if (!globals.ContainsKey(currentRequest.originalName)) throw new ArgumentNullException(); // TODO: need to throw a more valid exception
+
+                var currentResponse = new Response
                 {
-                    generics.Add(callable.Key);
-                }
+                    originalName = currentRequest.originalName,
+                    typeResolutions = currentRequest.typeResolutions,
+                    concreteCallable = globals[currentRequest.originalName].WithFullName(name => currentRequest.concreteName)
+                };
+
+                // Rewrite implementation
+                currentResponse = ReplaceTypeParamImplementationsSyntax.Apply(currentResponse);
+
+                // Rewrite calls
+                currentResponse = ReplaceTypeParamCallsSyntax.Apply(currentResponse, requests, responses);
+
+                responses.Add(currentResponse);
             }
 
-            var filter = new ResolveGenericsSyntax(new ScopeTransformation<ResolveGenericsExpression>(new ResolveGenericsExpression()));
+            var filter = new ResolveGenericsSyntax(responses
```

Comment:
There needs to be a test to make sure only what should be dropped is dropped (doesn't need to be part of this PR, but you should add that to your todo list).